staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go

```diff
@@ -513,3 +515,149 @@ func (u unionTransformers) TransformFromStorage(ctx context.Context, data []byte
 func (u unionTransformers) TransformToStorage(ctx context.Context, data []byte, dataCtx value.Context) (out []byte, err error) {
 	return u[0].TransformToStorage(ctx, data, dataCtx)
 }
+
+// DynamicTransformers holds transformers that are dynamically updated.
+type DynamicTransformers struct {
+	transformTracker *atomic.Value
+	SwapNotification *notificationChannel
+}
+
+type transformTracker struct {
+	transformerOverrides map[schema.GroupResource]value.Transformer
+	closeTransformers    context.CancelFunc
+	inFlight             *atomic.Int64
+	closing              *atomic.Bool
+}
+
+type notificationChannel struct {
+	once sync.Once
+	ch   chan struct{}
+}
+
+// Register registers a channel to be notified when the transformers are swapped.
+func (d *DynamicTransformers) Register() {
+	d.SwapNotification = &notificationChannel{
+		once: sync.Once{},
+		ch:   make(chan struct{}),
+	}
+}
+
+func (d *DynamicTransformers) Notified() <-chan struct{} {
+	return d.SwapNotification.ch
+}
+
+// NewDynamicTransformers initializes DynamicTransformers when api-server starts.
+func NewDynamicTransformers(
+	transformerOverrides map[schema.GroupResource]value.Transformer,
+	closeTransformers context.CancelFunc,
+) *DynamicTransformers {
+	dynamicTransformers := &DynamicTransformers{
+		transformTracker: &atomic.Value{},
+	}
+
+	tracker := &transformTracker{
+		transformerOverrides: transformerOverrides,
+		closeTransformers:    closeTransformers,
+		inFlight:             &atomic.Int64{},
+		closing:              &atomic.Bool{},
+	}
+	dynamicTransformers.transformTracker.Store(tracker)
+
+	return dynamicTransformers
+}
+
+// TransformerForResource returns the transformer for the given resource.
+func (d *DynamicTransformers) TransformerForResource(resource schema.GroupResource) value.Transformer {
+	return &resourceTransformer{
+		resource:         resource,
+		transformTracker: d.transformTracker,
+	}
+}
+
+// Set sets the transformer overrides. This method is not go routine safe and must only be called by the same, single caller throughout the lifetime of this object.
+func (d *DynamicTransformers) Set(
```

Comment:
if we plumb the max timeout of the set of transformers here, I think we can just wait a multiple of that timeout before closing the replaced transformers and simplify to drop the tracking

```go
Set(newTransformers, newTimeout) {
    oldTransformers, oldTimeout := swap(new, newTimeout)
    go func(){
        // give time for anything with a handle to the old transformers to begin an in-flight and complete with the timeout
        time.Sleep(... 2 * oldTimeout ...)
        // in-flights would now be complete or failed with timeout
        oldTransformers.close()
    }()
}
```

Reply:
implemented.